### DISCUSSION TITLE
[alist as a private cloud] new drvier - Rclone Crypt compatible Encryption Driver

### DISCUSSION THREAD

**Discussion by SeanHeuc** [AlistGo/alist#4788] [Show and tell]
Hi guys,

I implemented the a new driver: RcloneCrypt. it came from Rclone Project: https://rclone.org/crypt/  
it's working as transarent Encrytion/Decrytion Layer.

RcloneCrypt Storage:
![image](https://github.com/alist-org/alist/assets/866155/4636f1ed-9f95-41b0-bb0e-ce9b181d7df0)
Remote Storage(encrypted):
![image](https://github.com/alist-org/alist/assets/866155/5f264571-a94a-4f71-ab04-c1b4f0105350)
all data is available transparently to Alist
![image](https://github.com/alist-org/alist/assets/866155/9a1f6554-e5d5-4d40-aa9c-143214823a40)

the beta version is availble from
https://github.com/seanhe26/alist-private-cloud

I'll update the fork repo with roadmaps. the target is **Alist as a private cloud platform**

**I'm looking for volunteers to test and report bugs to me, pls reply this topic with your contact, (use base64 to protect your info)**

**Comment by SeanHeuc** [comment]
@Xhofe i did make some breaking change for the new driver, pls take a look if u have time. I think those changes are necessary, and will be helpful for other future features.

**Reply by xhofe** [reply]
LGTM, I'll take a look

**Reply by xhofe** [reply]
You could create a draft PR first so that more people could join the review

**Comment by anwen-anyi** [comment]
Do you need a foundation that can use Rclone?

I don't know how to fill in the parameters inside because I don't know how to use Rclone  :joy: :joy: :joy: :joy: :joy: 

![image](https://github.com/alist-org/alist/assets/56105412/955e1386-b179-4818-a159-8a85658341ff)

**Reply by SeanHeuc** [reply]
it's rclone crypt compatible, u can refer to their document on the parameter definition. this driver will work without rclone, both encrypt and decrypt

**Reply by anwen-anyi** [reply]
> it's rclone crypt compatible, u can refer to their document on the parameter definition. this driver will work without rclone, both encrypt and decrypt

I seem to understand some. I need to use RCLONE to encrypt locally, and then upload it to the driver. Is it used like this?
If that's the case, I don't seem to have rclone encrypted files

**Reply by SeanHeuc** [reply]
it's working as a transparent layer, u can directly upload to this driver's mount path, then file will be stored encrypted in the remote. u can download from driver's mount path, which is decrypted data from remote storage.   the concept is end-to-end encryption, u always have raw file while remote cloud storage only have encrypted data

**Reply by anwen-anyi** [reply]
Please check the private message of TG, I have left a message to you ~ @seanhe26 :joy_cat: :heart:

**Reply by SeanHeuc** [reply]
hi, i did not see ur reply on telegram, my id is  c2VhbjI2Ng==  (base64). do we have a dev/test group?

**Reply by anwen-anyi** [reply]
> hi, i did not see ur reply on telegram, my id is  c2VhbjI2Ng==  (base64). do we have a dev/test group?

Yes,in AList My Name is `安稳|有问题看文档你问的问题文档都有`

**Comment by luneew** [comment]
Looks fine and it works as a virtual driver layer.  However, what happens to the original unencrypted file and should it be filtered out or displayed transparently.  In addition, we should have a more general way to judge whether a file/folder is encrypted or not.

**Reply by SeanHeuc** [reply]
my personal idea is keep it as it is for the encrypted files in the remote storage. it gives people the flexibility  to be creative about the data. e.g. a group of people sharing files with the same password, they might want to copy/move/share the raw encrypted data.

same for identifing the encrypted folder/files, give people the freedom, it's also the how rclone deal with this.

**Comment by SeanHeuc** [comment]
some important fix is applied today, also a preview PR is submitted, now I could really use some help in testing.
PR: https://github.com/alist-org/alist/pull/4797
